File: Dynamic_Programming/1027. Longest Arithmetic Subsequence.cpp
/*
1027. Longest Arithmetic Subsequence
*/

// TC:O(n^2) SC:O(n^2)
class Solution {
public:
    int longestArithSeqLength(vector<int>& nums) {
        unordered_map<int, int> map;
        int offset = 500; // diff 負數 讓他變正數
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(1001,1));
        dp[0][0] = 1; // 只有一個數字 沒得選 差為0 len = 1
        
        int res = 1;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < i; j++){
                int diff = nums[i] - nums[j];
                dp[i][diff+offset] = dp[j][diff+offset] + 1;
                res = max(res, dp[i][diff+offset]);
               
            }
            
        }

        return res;
    }
};

/*
求等差的最長subsequeuce
n = 1000 可以接受n^2
每個點 i 跟他之前的人diff 以及其所能得到的最長長度 
dp[i][diff] = maxLen; 
*/

File: Dynamic_Programming/1027. Longest Arithmetic Subsequence.py
# DP TC:O(n^2) SC:O(n^2) 枚舉公差 TC:O(m * (m + n)) m為nums最大值 SC:O(m)
class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:
        '''
        求最長的等差數列 
        會讓人想到去沒舉可能的等差數列狀況
        數字大小從0-500 所以最大的等差只會是500 可能是-500 ~ 500
        '''
        # 枚舉公差
        ans = 0
        m = max(nums)
        for d in range(-m, m + 1):
            f = [0] * (m + 1)
            for x in nums:
                f[x] = f[x - d] + 1 if 0 <= x - d <= m else 1
            ans = max(ans, max(f))
        return ans

        '''
        遞推
        f = [{} for _ in range(len(nums))]
        for i, x in enumerate(nums):
            for j in range(i - 1, -1, -1):
                d = x - nums[j]  # 公差
                if d not in f[i]:
                    f[i][d] = f[j].get(d, 1) + 1
        return max(max(d.values()) for d in f[1:])
        '''

        '''
        遞歸
        @cache
        def dfs(i:int) -> dict[int,int]:
            max_len = {}
            for j in range(i-1, -1, -1):
                d = nums[i] - nums[j]
                if d not in max_len:
                    max_len[d] = dfs(j).get(d,1) + 1
            return max_len
        return max(max(dfs(i).values()) for i in range(1, len(nums)))
        '''
